### DISCUSSION TITLE
Consider unacknowledged messages for delete queue if-empty

### DISCUSSION THREAD

**Discussion by zvirja** [rabbitmq/rabbitmq-server#10725] [Other]
### Is your feature request related to a problem? Please describe.

In our product we have jobs subsystem where we create a queue per job. The idea is that queue contains job items and queue consumer handles those job items. As a result of single item handling we could enqueue more job items to the end of the queue. So queue could dynamically grow and shrink. But eventually queue becomes empty when all the items are handled.

The problem is to detect a moment when the queue is empty and it's safe to finalize the job and clean up the queue. We want to do that as quick as possible, so job is marked as done fast. It would be very convenient to use `if-empty` parameter for the `queue delete` action (so if we managed to successfully delete the queue we know that we are done), but unfortunately `if-empty` does not consider non-acked messages. Apparently it was intentional: [see this mail discussion](https://lists.rabbitmq.com/pipermail/rabbitmq-discuss/2013-January/024735.html).

### Describe the solution you'd like

Either would work for my problem:
- Have something like `if-empty` for `Delete Queue`, but also consider the non-ACKed messages
- Allow to get up-to-date queue stats to detect if queue is empty _including non-acknowledged_. Currently `Declare Queue` returns only Ready count, so it's not suitable.

### Describe alternatives you've considered

For now we are using Management API, but that is non-reliable, slow and extremely inaccurate. Sometimes it shows zero count, while in reality the queue is not empty. Often it out of date. So the only desperate thing we are doing is querying stats from Management API for a few times with a delay. If every time queue is empty - we believe it's empty. But that approach is slow, clumsy and still does not guarantee the correctness. We also delay job completion for ~30 secs which degrades user experience.

### Additional context

I would say that the existing `if-empty` behavior for delete queue is not very practical. When you have messages in flight (non-acked yet), it is still possible to that they are NACK-ed (e.g. due to connection issues) and they are returned back to queue. So you actually never want to delete a queue if you have those, as that's a possible to get them lost. That's exactly what happened when I tested them locally and app crashed after deleting the queue but before ACK-ing the messages in flight - messages simply got lost.

The main argument for the current behavior [given the mail](https://lists.rabbitmq.com/pipermail/rabbitmq-discuss/2013-January/024735.html) was alignment and consistency across various API. I would though argue that the statement from the mail is true:

>  Also, conceptually, messages that are pending acknowledgement are not really in the queue anymore; or, rather, there is nothing in AMQP that suggests, let alone mandates, such a view.

Given that they are returned to the queue in case of lost connection makes them still belonging to the queue.

I would suggest to just change the existing behavior of `if-empty` to consider non-ACKed or add a configuration flag for that.

**Comment by michaelklishin** [comment]
This was discussed several times and was discussed against.

This violates two fairly common assumptions:

 * the idea that a delivered and unacknowledged messages are "owned" by the application (an assumption that exists in all messaging protocols RabbitMQ supports)
 * the idea that an unacknowledged message can be acknowledged at any moment after it's been delivered (and before consumer delivery timeout hits, or connection closes, various failure-like scenarios like that)

If you really want to delete a queue, just delete it without any conditions. Otherwise, a queue with oustanding unacknowledged messages is not really semantically empty.

**Reply by zvirja** [reply]
@michaelklishin Thanks for a prompt reply!

Maybe I just don't see something, but I simply cannot how one could create a practical robust design with the ideas you mentioned above.

I would like to have robustness, so that messages are not lost and are retried if application crashes or times out. Currently that is guaranteed by NACK-ing it and putting back to queue. So if message is not ACK-ed, we have a guarantee that it will be delivered again, until we manage to handle it. That is awesome.

That however is not the case if you delete the queue while having non-ACKed messages, as then the robustness above is lost. The only solution I see is to not remove the queue until we really ACK all the messages.

You suggest that application "owns" messages which are unacknowledged yet. But I simply don't see how application can _realistically_ implement robustness and guarantee that message is retried if e.g. app crashes or is forcibly killed by OS. The only approach would be introduce another kind of persistence to track a matter of completion, but that would be just re-implementing what RabbitMQ is doing today. And even that would not work, as app could be crashed in a moment when message is delivered to the app, but before event handler even has a chance to start execution.

Given that you discussed that internally and rejected idea of considering unacknowledged items during queue deletion, how did you envision solving the robustness issue? Could be that I simply am unaware of something. What was the suggested _realistic_ design to make sure that the messages are not lost?

Thank you!

**Reply by lukebakken** [reply]
> I would like to have robustness, so that messages are not lost and are retried if application crashes or times out

Don't delete the queue.

**Reply by zvirja** [reply]
If you read the original scenario I have, the issue is not necessarily around deletion of the queue. That could indeed be done later. The issue is detecting that queue becomes fully empty, so that I know that the job is finished. I have a distributed application with horizontal scaling, so I have many nodes and synchronizing across nodes is problematic. How can I tell that queue is truly empty using the API we have today? All the existing AMQP protocol methods are simply ignoring non-ACKed messages.

Thanks.

**Reply by lukebakken** [reply]
> If you read the original scenario I have, the issue is not necessarily around deletion of the queue. That could indeed be done later. The issue is detecting that queue becomes fully empty, so that I know that the job is finished. 

Yes, I read your original scenario. Don't delete queues if you can't be certain they are empty given what RabbitMQ provides. A better solution would be to have your job processing application publish a message indicating that it is finished. That way you don't have to depend on RabbitMQ's internal state.

You're asking for a change in long-established behavior. It's unlikely we would consider a change such as this unless a major, paying user of RabbitMQ suggested it, and even then, it would be only after much planning and discussion.

**Reply by zvirja** [reply]
> A better solution would be to have your job processing application publish a message indicating that it is finished. That way you don't have to depend on RabbitMQ's internal state.

The nature of the jobs I implement is that I dynamically generate new job entries on flight. So I simply don't know how many of them I am going to have and the only way to detect that I am done is actually to detect that the queue is empty. There are no other realistic ways, unless I start doing counting on a side. And even that would not work reliably for cases with dead letters and etc.

So for my design I desperately need some API to tell if queue is truly empty and if RabbitMQ could tell me that - it would work like a charm.

> You're asking for a change in long-established behavior. It's unlikely we would consider a change such as this unless a major, paying user of RabbitMQ suggested it, and even then, it would be only after much planning and discussion.

We don't necessarily have to change the existing behavior. The answer could be to introduce a new API to get queue size including non-ACKed messages. As clearly there is a use case for that and I believe that our app is not the only one needing it. In this case the change will be non-disruptive. Is that something which is more realistic to do?

**Reply by lukebakken** [reply]
> The answer could be to introduce a new API to get queue size including non-ACKed messages

Good point. Feel free to open a PR to contribute such a feature.

**Reply by zvirja** [reply]
Unfortunately, I don't know Erlang. I don't know RabbitMQ either to make such deep changes, like designing and introducing new AMQP protocol API. Is that something which is possible raise as a feature request to the core team? So that maintainers could spec the design and implement the change?

I could create such a Feature Request. I believe it's a feature many will benefit from. Do you think it's realistic something like this will be implemented?

**Reply by lukebakken** [reply]
> Is that something which is possible raise as a feature request to the core team? So that maintainers could spec the design and implement the change?

Both @michaelklishin and I are on the core team.

> I believe it's a feature many will benefit from.

It would be great if other RabbitMQ users could request this feature, especially if they pay for RabbitMQ (via a support contract). They should follow up here to express interest.

> Do you think it's realistic something like this will be implemented?

If a customer requests it, it is much more likely. Otherwise, Team RabbitMQ has a long, long list of higher priority work to do.

**Reply by michaelklishin** [reply]
Alternatively you can consider setting queue TTL on your queues, not any additional features or behaviour changes. And maybe message TTL but that's a much more consequential option to consider.

As long as the consumers are there, the queue won't be deleted. When they go away, eventually they will be.

A queue with delivered messages pending acknowledgement are not empty. We are not interested in expanding the matrix of features related to queue deletion, there are more than enough.

**Reply by zvirja** [reply]
Okay, so it seems that the plan is to keep it open and then see if there is more interesting. Gotcha!

> If a customer requests it, it is much more likely.

@lukebakken What is that? Like I am also a user of the product 😄 Do you mean those paid requests/support contracts?

@michaelklishin Thanks for the advice. I considered using that as well, but I don't see how to make it work. Queue TTL assumes that there shouldn't be any consumers. But specifically in my case I have chicken-and-egg problem. I need to somehow detect that queue became totally empty, so I could unsubscribe from it, so it could be removed. But I don't have any API to check if the queue is actually empty to accomplish that...

> We are not interested in expanding the matrix of features related to queue deletion, there are more than enough.

That is fair enough. More I am thinking about it, more it makes sense to have feature about total queue size (Ready + non-ACK), as then it could be used here.

---

@michaelklishin @lukebakken I have another question though, out of curiosity. With the current implementation and from what we have discussed above, if you delete queue with pending non-ACKed messages, you effectively have data loss (e.g. if consumer crashes). With that it means that `QueueDelete(if-empty: true)` with active consumers is ALWAYS going to have problems of data loss. Like ALWAYS. And those issues are quite subtle to catch.

If QueueDelete checked for _total_ count, then it would be another story. If combined with `Publish(mandatory: true)`, you would technically prevent _any_ possible data loss. Like if deletion of the queue fails - it means that it's not safe yet to remove it. If it managed to remove it, then publishing would fail. It would be caught at both sides. The key here would be atomicity of "transactions", so usage of `QueueDelete(if-empty: true)` would be a must.

What is the envisioned use case of `QueueDelete(if-empty: true)` in its current behavior? Like why people would ever use it and what for? Which practical problem does it solve which cannot be solved with e.g. `QueueDeclare(passive: true)->Count + QueueDelete())`?

Would be happy if you elaborate on that in more detail, so I and hopefully others will have answers 🙏

**Reply by lukebakken** [reply]
> Do you mean those paid requests/support contracts?

Yes. Feature requests from people who use RabbitMQ for free are rarely taken into consideration, unless the request is accompanied by a PR that implements it.

> if you delete queue with pending non-ACKed messages, you effectively have data loss (e.g. if consumer crashes). With that it means that QueueDelete(if-empty: true) with active consumers is ALWAYS going to have problems of data loss. Like ALWAYS. And those issues are quite subtle to catch.

Yep. And yet, we don't receive reports about this issue 🤷‍♂️ 

> QueueDeclare(passive: true)->Count + QueueDelete())?

Those operations would not be atomic.

You make a good case for changing the `QueueDelete / if-empty` behavior, or adding a different API that takes outstanding confirms into account. Again, if this were to be requested from a large number of RabbitMQ users, or from a paying customer, we would raise its priority.

As it stands, Matthias' response [here](https://lists.rabbitmq.com/pipermail/rabbitmq-discuss/2013-January/024735.html) _still applies today_:

> Alas I don't  know how we'd go about meeting such a requirement; I'd be reluctant to change the semantics of something that has been around for five years, particularly since we've had zero complaints to date about the current semantics.

**Reply by michaelklishin** [reply]
Consumers are usually responsible for declaring and deleting their own queues. Deleting someone else's queues can result in data loss in many different situations. In modern messaging protocols, applications are responsible for the topology. RabbitMQ cannot know if what an application attempts to do is safe or not. Saving applications from themselves has been an ongoing topic over the last ≈ 15 years or so that I've been a contributor, and it's a battle you cannot win.

Queue and message TTL give you more options. So do auto-delete and exclusive queues. For example, another application simply cannot delete a queue declared on a different connections (another app, a different app instance). If your apps cannot agree on when and how to delete queues, use options that do not require explicit deletion.

If you need a shared data structure where consumers do not or cannot coordinate, consider using a stream. Streams are rarely deleted and their only way of deleting data is TTL.

There are many drastically more important problems to solve in RabbitMQ, its client libraries, its operational characteristics, and so on.

Introducing a protocol extension (a new protocol method to obtain more queue metrics) would mean updating at least several client libraries to start with. That can roughly triple or quadruple the amount of work.
